Improve performance of aromaticity detection for large molecules

d-b-w opened the pull request:
For very large molecules, the repeated counting of total number
of ring atoms is a bottleneck in aromaticity detection! This
pulls that total count out of the function that is checking
aromaticity of a particular ring set.

For an example molecule (PDB ID: 3EOH), this basically
removes aromaticity detection from a performance profile
of sanitization. (on my laptop, makes sanitization about
1.5s faster for 3EOH).

My only hesitation with this commit is that the check
that requires knowledge of the total number of ring
atoms may not be needed at all?

d-b-w commented:
I can reproduce the test failures, and agree that for the molecule that you mention this makes sense. Let me read through the test to understand what I can, and I'll post and update (and change the title and description of this PR).

greglandrum commented:
> I can reproduce the test failures, and agree that for the molecule that you mention this makes sense. Let me read through the test to understand what I can, and I'll post and update (and change the title and description of this PR).

I'm not sure the title should change. I do think it's worth adding an explicit test for #3256. If it helps, I can do a PR for that (and the fix to the MQN tests)

greglandrum commented:
@d-b-w how do you want to proceed with this?

d-b-w commented:
I'm happy to address this - I just had a couple of other things on my plate last week.

I understand why the old code was problematic now - it was a check for done atoms, rather than done bonds.

d-b-w commented:
I've updated the commit to completely remove the "done atoms" check in Aromaticity.cpp. This addresses the performance and accuracy issues discussed in this case. I also updated the tests, and added a new test in catch_graphmol.cpp to explicitly actively check these systems for correct envelope aromaticity.

I did notice one odd thing... For the test that I had to update - the line in aromat_regress.txt says: 

     N12C(N3N=C(C)C=C(N4C1=CC(C)=N4)3)=C(C=O)C(=N2)C	Cc1cc2n(n1)c1cc(C)nn1c1c(C=O)c(C)nn21	15	[0, 1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 13, 14, 17, 18]

But before this change, if I render this mol back as SMILES I get:

    Cc1cc2n(n1)-c1cc(C)nn1-c1c(C=O)c(C)nn1-2

(note the explicit single bonds between aromatic atoms). Was adding single bonds between aromatic atoms a new fix?

greglandrum commented:
Hi @d-b-w : the CI failures are real ones this time.

d-b-w commented:
ugh! I originally wrote the test in Python and then got sloppy when translating. Sorry about that.

The current failure is in a java test. The same test passes on Mac. It looks like this is the assertion that fails:

* https://github.com/rdkit/rdkit/blob/master/Code/JavaWrappers/gmwrapper/src-test/org/RDKit/DiversityPickerTests.java#L163

Does that seem related to this?

d-b-w commented:
oooook. I had to kill this after a minute or two runtime on a nanotube system. Looks like we need _some_ sort of kill switch.

When I next have a chance, I'll play with using _bonds_ that are complete rather than atoms.

greglandrum commented:
> ugh! I originally wrote the test in Python and then got sloppy when translating. Sorry about that.
> 
> The current failure is in a java test. The same test passes on Mac. It looks like this is the assertion that fails:
> 

Failures in that test are generally not real. The test has a random component to it (intentionally) and that occasionally results in a failure. I restarted that failed build on Azure DevOps

greglandrum commented on rdkit/Chem/UnitTestDescriptors.py:
```diff
@@ -84,25 +84,19 @@ def testMolFormula(self):
   def testMQNDetails(self):
-    refFile = os.path.join(RDConfig.RDCodeDir, 'Chem', 'test_data', 'MQNs_regress.pkl')
-    refFile2 = os.path.join(RDConfig.RDCodeDir, 'Chem', 'test_data', 'MQNs_non_strict_regress.pkl')
+    refFile = os.path.join(os.path.dirname(__file__), 'test_data', 'MQNs_regress.pkl')
```
Any particular reason you made these changes? They aren't wrong, but the pattern of using `RDConfig` is pretty common in the python tests

d-b-w replied:
oh - good question. I usually build via installation into a Schrödinger suite, because there are things that I want to test there. These test files don't get installed.

greglandrum replied:
I *think* (and you should verify this) that if you run cmake with `-DRDK_INSTALL_PYTHON_TESTS=ON`  the python tests and data will be installed.

d-b-w commented:
I think it would be good to hold off on the merge until I figure out whether this causes a hang in a system that ran in finite time before.

greglandrum commented:
@d-b-w : any news here?

greglandrum commented on Code/GraphMol/Aromaticity.cpp:
```diff
@@ -361,17 +366,27 @@ void applyHuckelToFused(
   INT_VECT comb;
   pos = -1;
+
+  size_t nRingBonds;
```
wouldn't it be a lot more efficient to do this with a dynamic_bitset to track which bonds have been seen and then calling count() at the end?

d-b-w replied:
ah, probably true. I'll update.

d-b-w replied:
Although setting and getting values in a bitset is not so so fast, because of all the bitmasks. Looks like the unordered_set approach is about 1.2x slower for this portion of the call.

d-b-w commented:
I think that the test failures are all in doctest and are unrelated:

    Exception occurred:
      File "/usr/share/miniconda/envs/rdkit_build/lib/python3.7/typing.py", line 969, in get_type_hints
        base_globals = sys.modules[base.__module__].__dict__
    KeyError: 'Boost.Python'

 I'll merge master, because I think that there were some changes to the docs, and try again.

d-b-w commented:
I've done the testing that I need to on this to feel good about it.

greglandrum commented:
> I think that the test failures are all in doctest and are unrelated:
> 
> ```
> Exception occurred:
>   File "/usr/share/miniconda/envs/rdkit_build/lib/python3.7/typing.py", line 969, in get_type_hints
>     base_globals = sys.modules[base.__module__].__dict__
> KeyError: 'Boost.Python'
> ```
> 
> I'll merge master, because I think that there were some changes to the docs, and try again.

yeah, those failures were due to a problem with sphinx 3.2. The changes to master cleared them up

greglandrum commented:
Thanks @d-b-w !
